File: 回測模組/strategy_modules/data_processor.py
"""
資料處理模組
負責載入和處理回測資料
"""

import os
import logging
import pandas as pd
import numpy as np
from datetime import datetime
from typing import Dict, Tuple, Optional
from decimal import Decimal

logger = logging.getLogger(__name__)


class DataProcessor:
    """資料處理器"""

    def __init__(self, config: Dict):
        """
        初始化資料處理器

        Args:
            config: 設定字典
        """
        self.config = config
        self.company_info_cache = None  # 快取公司資訊

    def load_feature_data(self, stock_id: str, date: str) -> pd.DataFrame:
        """
        載入特徵資料

        Args:
            stock_id: 股票代碼
            date: 日期 (YYYY-MM-DD)

        Returns:
            特徵資料 DataFrame
        """
        feature_path = f'D:/feature_data/feature/{date}/{stock_id}.parquet'

        if not os.path.exists(feature_path):
            logger.error(f"檔案不存在: {feature_path}")
            return pd.DataFrame()

        logger.info(f"載入資料: {feature_path}")
        df = pd.read_parquet(feature_path)

        # 轉換時間欄位
        if 'time' in df.columns:
            df['time'] = pd.to_datetime(df['time'])
        else:
            logger.error("資料中沒有 'time' 欄位")
            return pd.DataFrame()

        # 篩選交易時段 (09:00-13:30)
        df = df[
            (df['time'].dt.time >= pd.Timestamp('09:00:00').time()) &
            (df['time'].dt.time <= pd.Timestamp('13:30:00').time())
        ]

        logger.info(f"載入 {len(df)} 筆資料")
        return df

    def process_trade_data(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        處理交易資料

        Args:
            df: 原始資料 DataFrame

        Returns:
            處理後的交易資料 DataFrame
        """
        if 'type' not in df.columns:
            logger.warning("資料中沒有 'type' 欄位，假設全部為 Trade 資料")
            return df

        # 分離 Trade 和 Depth 資料
        trade_df = df[df['type'] == 'Trade'].copy()
        depth_df = df[df['type'] == 'Depth'].copy()

        logger.info(f"Trade 資料筆數: {len(trade_df)}")
        logger.info(f"Depth 資料筆數: {len(depth_df)}")

        if len(trade_df) == 0:
            logger.warning("沒有 Trade 資料")
            return pd.DataFrame()

        # 將 Trade 資料中的掛單欄位值從 0 改為 NaN（這樣才能被 ffill 填充）
        order_book_columns = [
            'bid1_volume', 'bid2_volume', 'bid3_volume', 'bid4_volume', 'bid5_volume',
            'ask1_volume', 'ask2_volume', 'ask3_volume', 'ask4_volume', 'ask5_volume',
            'bid_volume_5level', 'ask_volume_5level', 'bid_ask_ratio'
        ]

        for col in order_book_columns:
            if col in trade_df.columns:
                trade_df[col] = trade_df[col].replace(0, np.nan)

        # 合併資料，保留所有欄位
        merged_df = pd.concat([trade_df, depth_df]).sort_values('time')

        # 向前填充掛單資料欄位
        for col in order_book_columns:
            if col in merged_df.columns:
                merged_df[col] = merged_df[col].ffill()

        # 只保留 Trade 類型的資料進行交易邏輯判斷
        result_df = merged_df[merged_df['type'] == 'Trade'].copy()

        # 預先補齊 5 檔合計，避免逐筆計算成本
        bid_cols = [f"bid{i}_volume" for i in range(1, 6)]
        ask_cols = [f"ask{i}_volume" for i in range(1, 6)]

        if all(col in result_df.columns for col in bid_cols):
            if 'bid_volume_5level' not in result_df.columns:
                result_df['bid_volume_5level'] = result_df[bid_cols].sum(axis=1, min_count=1)
            else:
                mask = result_df['bid_volume_5level'].isna() | (result_df['bid_volume_5level'] == 0)
                if mask.any():
                    result_df.loc[mask, 'bid_volume_5level'] = result_df.loc[mask, bid_cols].sum(axis=1, min_count=1)

        if all(col in result_df.columns for col in ask_cols):
            if 'ask_volume_5level' not in result_df.columns:
                result_df['ask_volume_5level'] = result_df[ask_cols].sum(axis=1, min_count=1)
            else:
                mask = result_df['ask_volume_5level'].isna() | (result_df['ask_volume_5level'] == 0)
                if mask.any():
                    result_df.loc[mask, 'ask_volume_5level'] = result_df.loc[mask, ask_cols].sum(axis=1, min_count=1)

        logger.info(f"處理後 Trade 資料筆數: {len(result_df)}")
        return result_df

    def merge_depth_data(self, trade_df: pd.DataFrame, depth_df: pd.DataFrame) -> pd.DataFrame:
        """
        合併深度資料到交易資料

        Args:
            trade_df: 交易資料 DataFrame
            depth_df: 深度資料 DataFrame

        Returns:
            合併後的 DataFrame
        """
        order_book_columns = [
            'bid1_volume', 'bid2_volume', 'bid3_volume', 'bid4_volume', 'bid5_volume',
            'ask1_volume', 'ask2_volume', 'ask3_volume', 'ask4_volume', 'ask5_volume',
            'bid_volume_5level', 'ask_volume_5level', 'bid_ask_ratio'
        ]

        # 將 Trade 資料中的掛單欄位值從 0 改為 NaN
        for col in order_book_columns:
            if col in trade_df.columns:
                trade_df[col] = trade_df[col].replace(0, np.nan)

        # 合併資料
        merged_df = pd.concat([trade_df, depth_df]).sort_values('time')

        # 向前填充掛單資料
        for col in order_book_columns:
            if col in merged_df.columns:
                merged_df[col] = merged_df[col].ffill()

        # 只保留 Trade 類型的資料
        result_df = merged_df[merged_df['type'] == 'Trade'].copy()

        # 預先補齊 5 檔合計，避免逐筆計算成本
        bid_cols = [f"bid{i}_volume" for i in range(1, 6)]
        ask_cols = [f"ask{i}_volume" for i in range(1, 6)]

        if all(col in result_df.columns for col in bid_cols):
            if 'bid_volume_5level' not in result_df.columns:
                result_df['bid_volume_5level'] = result_df[bid_cols].sum(axis=1, min_count=1)
            else:
                mask = result_df['bid_volume_5level'].isna() | (result_df['bid_volume_5level'] == 0)
                if mask.any():
                    result_df.loc[mask, 'bid_volume_5level'] = result_df.loc[mask, bid_cols].sum(axis=1, min_count=1)

        if all(col in result_df.columns for col in ask_cols):
            if 'ask_volume_5level' not in result_df.columns:
                result_df['ask_volume_5level'] = result_df[ask_cols].sum(axis=1, min_count=1)
            else:
                mask = result_df['ask_volume_5level'].isna() | (result_df['ask_volume_5level'] == 0)
                if mask.any():
                    result_df.loc[mask, 'ask_volume_5level'] = result_df.loc[mask, ask_cols].sum(axis=1, min_count=1)

        return result_df

    def get_reference_price(self, stock_id: str, date: str, close_path: str = None) -> float:
        """
        取得參考價格（昨收價）

        Args:
            stock_id: 股票代碼
            date: 日期 (YYYY-MM-DD)
            close_path: 收盤價檔案路徑

        Returns:
            參考價格
        """
        if close_path is None:
            close_path = r'C:\Users\User\Documents\_02_bt\Backtest_tick_module\close.parquet'

        ref_price = None

        if os.path.exists(close_path):
            try:
                close_df = pd.read_parquet(close_path)

                if stock_id in close_df.columns:
                    # 將日期格式轉換
                    close_df.index = pd.to_datetime(close_df.index)
                    current_date = pd.to_datetime(date)

                    # 找到小於當前日期的最後一個交易日
                    prev_dates = close_df.index[close_df.index < current_date]
                    if len(prev_dates) > 0:
                        prev_date = prev_dates[-1]
                        ref_price = close_df.loc[prev_date, stock_id]
                        logger.info(f"從 close.parquet 讀取到昨收價: {ref_price:.2f} (日期: {prev_date.date()})")
                    else:
                        logger.warning(f"在 close.parquet 中找不到 {date} 之前的收盤價")
                else:
                    logger.warning(f"在 close.parquet 中找不到股票 {stock_id}")
            except Exception as e:
                logger.error(f"讀取 close.parquet 失敗: {e}")
        else:
            logger.warning(f"找不到 close.parquet 檔案: {close_path}")

        return ref_price

    @staticmethod
    def get_tick_size_decimal(price: float) -> Decimal:
        """
        根據價格回傳對應的升降單位

        Args:
            price: 價格

        Returns:
            升降單位（Decimal）
        """
        if pd.isna(price):
            return Decimal('0')

        p = float(price)

        tick_rules = [
            (1000, 5), (500, 1), (100, 0.5), (50, 0.1), (10, 0.05)
        ]

        for threshold, tick in tick_rules:
            if p >= threshold:
                return Decimal(str(tick))

        return Decimal('0.01')

    @staticmethod
    def calculate_limit_up(previous_close: float) -> float:
        """
        計算漲停板：(昨收 * 1.1) 根據該價位 Tick 向下取整

        Args:
            previous_close: 昨收價

        Returns:
            漲停價
        """
        if pd.isna(previous_close):
            return 0.0

        raw_limit = Decimal(str(previous_close)) * Decimal('1.10')
        tick = DataProcessor.get_tick_size_decimal(float(raw_limit))
        limit_up = (raw_limit // tick) * tick

        return round(float(limit_up), 2)

    @staticmethod
    def calculate_limit_down(previous_close: float) -> float:
        """
        計算跌停板：(昨收 * 0.9) 根據該價位 Tick 向上取整

        Args:
            previous_close: 昨收價

        Returns:
            跌停價
        """
        if pd.isna(previous_close):
            return 0.0

        raw_limit = Decimal(str(previous_close)) * Decimal('0.90')
        tick = DataProcessor.get_tick_size_decimal(float(raw_limit))
        # 向上取整
        limit_down = (((raw_limit + tick - Decimal('0.01')) // tick) * tick).normalize()

        return float(limit_down)

    def add_calculated_columns(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        添加計算欄位

        Args:
            df: 原始資料 DataFrame

        Returns:
            添加計算欄位後的 DataFrame
        """
        # 添加空欄位，稍後會在主程式中填充
        df['day_high_growth_rate'] = 0.0
        df['bid_avg_volume'] = 0.0
        df['ask_avg_volume'] = 0.0
        df['balance_ratio'] = 0.0
        df['day_high_breakout'] = False

        return df

    def get_company_name(self, stock_id: str) -> str:
        """
        取得公司名稱

        Args:
            stock_id: 股票代碼

        Returns:
            公司名稱，如果找不到則返回股票代碼
        """
        try:
            # 如果尚未載入公司資訊，嘗試載入
            if self.company_info_cache is None:
                # 嘗試從不同路徑載入公司資訊
                possible_paths = [
                    'company_basic_info.parquet',
                    'D:/feature_data/company_basic_info.parquet',
                    'C:/Users/User/Documents/_02_bt/Backtest_tick_module/company_basic_info.parquet'
                ]

                for path in possible_paths:
                    if os.path.exists(path):
                        self.company_info_cache = pd.read_parquet(path)
                        logger.info(f"載入公司資訊從: {path}")
                        break

                # 如果有成功載入，建立名稱對應字典
                if self.company_info_cache is not None:
                    if '公司簡稱' in self.company_info_cache.columns and 'stock_id' in self.company_info_cache.columns:
                        self.name_map = self.company_info_cache.set_index('stock_id')['公司簡稱'].to_dict()
                    else:
                        logger.warning("公司資訊檔案格式不正確")
                        self.name_map = {}
                else:
                    logger.warning("找不到公司資訊檔案")
                    self.name_map = {}

            # 返回公司名稱
            return self.name_map.get(stock_id, stock_id)

        except Exception as e:
            logger.warning(f"載入公司名稱失敗: {e}")
            return stock_id

    def filter_trading_hours(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        篩選交易時段資料

        Args:
            df: 原始資料 DataFrame

        Returns:
            篩選後的 DataFrame
        """
        # 篩選交易時段 (09:00-13:30)
        trading_df = df[
            (df['time'].dt.time >= pd.Timestamp('09:00:00').time()) &
            (df['time'].dt.time <= pd.Timestamp('13:30:00').time())
        ].copy()

        return trading_df

    def validate_data(self, df: pd.DataFrame) -> bool:
        """
        驗證資料完整性

        Args:
            df: 資料 DataFrame

        Returns:
            是否通過驗證
        """
        # 檢查必要欄位
        required_columns = ['time', 'price', 'volume', 'tick_type']
        missing_columns = [col for col in required_columns if col not in df.columns]

        if missing_columns:
            logger.error(f"缺少必要欄位: {missing_columns}")
            return False

        # 檢查資料數量
        if len(df) == 0:
            logger.error("資料為空")
            return False

        # 檢查時間順序
        if not df['time'].is_monotonic_increasing:
            logger.warning("資料時間順序不正確，將進行排序")
            df.sort_values('time', inplace=True)

        return True

File: 回測模組/CSharp/BacktestModule/Strategy/DataProcessor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using BacktestModule.Core.Models;

namespace BacktestModule.Strategy
{
    /// <summary>
    /// Data processor: loads parquet feature data, processes trade/depth data,
    /// calculates tick sizes and limit prices.
    /// Note: Parquet reading requires Apache.Arrow or Parquet.Net NuGet package.
    /// This implementation provides the structural skeleton; actual parquet I/O
    /// uses placeholder methods that should be backed by a concrete parquet reader.
    /// </summary>
    public class DataProcessor
    {
        private readonly Dictionary<string, object> _config;
        private Dictionary<string, string> _companyNameCache;

        public DataProcessor(Dictionary<string, object> config)
        {
            _config = config;
            _companyNameCache = null;
        }

        /// <summary>
        /// Loads feature data from parquet file for a given stock and date.
        /// Path: D:/feature_data/feature/{date}/{stockId}.parquet
        /// Converts time column, filters 09:00-13:30.
        /// </summary>
        public List<TickData> LoadFeatureData(string stockId, string date)
        {
            string featurePath = $@"D:\feature_data\feature\{date}\{stockId}.parquet";

            if (!File.Exists(featurePath))
            {
                Console.WriteLine($"[ERROR] File not found: {featurePath}");
                return new List<TickData>();
            }

            Console.WriteLine($"[INFO] Loading data: {featurePath}");

            // Load from parquet (requires Parquet.Net or Apache.Arrow)
            var data = ParquetHelper.ReadParquetToTickData(featurePath);

            // Filter trading hours: 09:00 - 13:30
            var tradingStart = new TimeSpan(9, 0, 0);
            var tradingEnd = new TimeSpan(13, 30, 0);
            data = data.Where(t => t.Time.TimeOfDay >= tradingStart && t.Time.TimeOfDay <= tradingEnd).ToList();

            Console.WriteLine($"[INFO] Loaded {data.Count} records");
            return data;
        }

        /// <summary>
        /// Processes trade data: separates Trade/Depth, forward-fills order book columns,
        /// keeps only Trade rows, computes 5-level bid/ask totals.
        /// </summary>
        public List<TickData> ProcessTradeData(List<TickData> rawData)
        {
            if (rawData == null || rawData.Count == 0)
                return new List<TickData>();

            // Check if 'type' column exists meaningfully
            bool hasTypeColumn = rawData.Any(r => !string.IsNullOrEmpty(r.Type));
            if (!hasTypeColumn)
            {
                Console.WriteLine("[WARNING] No 'type' column found, assuming all data is Trade.");
                return rawData;
            }

            var tradeRows = rawData.Where(r => r.Type == "Trade").ToList();
            var depthRows = rawData.Where(r => r.Type == "Depth").ToList();

            Console.WriteLine($"[INFO] Trade rows: {tradeRows.Count}");
            Console.WriteLine($"[INFO] Depth rows: {depthRows.Count}");

            if (tradeRows.Count == 0)
            {
                Console.WriteLine("[WARNING] No Trade data.");
                return new List<TickData>();
            }

            // Set order book columns in Trade rows to NaN (represented as 0 -> NaN)
            foreach (var row in tradeRows)
            {
                SetOrderBookToNaN(row);
            }

            // Merge trade + depth, sort by time
            var merged = new List<TickData>(tradeRows.Count + depthRows.Count);
            merged.AddRange(tradeRows);
            merged.AddRange(depthRows);
            merged = merged.OrderBy(t => t.Time).ToList();

            // Forward-fill order book columns
            ForwardFillOrderBook(merged);

            // Keep only Trade rows
            var result = merged.Where(r => r.Type == "Trade").ToList();

            // Compute bid_volume_5level and ask_volume_5level if missing
            foreach (var row in result)
            {
                EnsureFiveLevelTotals(row);
            }

            Console.WriteLine($"[INFO] Processed Trade rows: {result.Count}");
            return result;
        }

        /// <summary>
        /// Alternative merge method (same logic as ProcessTradeData).
        /// </summary>
        public List<TickData> MergeDepthData(List<TickData> tradeData, List<TickData> depthData)
        {
            foreach (var row in tradeData)
                SetOrderBookToNaN(row);

            var merged = new List<TickData>(tradeData.Count + depthData.Count);
            merged.AddRange(tradeData);
            merged.AddRange(depthData);
            merged = merged.OrderBy(t => t.Time).ToList();

            ForwardFillOrderBook(merged);

            var result = merged.Where(r => r.Type == "Trade").ToList();
            foreach (var row in result)
                EnsureFiveLevelTotals(row);

            return result;
        }

        /// <summary>
        /// Gets the previous trading day's close price for a stock.
        /// </summary>
        public double? GetReferencePrice(string stockId, string date, string closePath = null)
        {
            if (closePath == null)
                closePath = @"C:\Users\User\Documents\_02_bt\Backtest_tick_module\close.parquet";

            if (!File.Exists(closePath))
            {
                Console.WriteLine($"[WARNING] Close price file not found: {closePath}");
                return null;
            }

            try
            {
                // Read close.parquet: index=dates, columns=stock_ids, values=close prices
                var closeData = ParquetHelper.ReadCloseParquet(closePath);
                var currentDate = DateTime.Parse(date);

                // Find the last trading day before currentDate
                var prevDates = closeData.Keys.Where(d => d < currentDate).OrderBy(d => d).ToList();
                if (prevDates.Count == 0)
                {
                    Console.WriteLine($"[WARNING] No previous close price found for {date}");
                    return null;
                }

                var prevDate = prevDates.Last();
                if (closeData[prevDate].TryGetValue(stockId, out double refPrice))
                {
                    Console.WriteLine($"[INFO] Previous close: {refPrice:F2} (date: {prevDate:yyyy-MM-dd})");
                    return refPrice;
                }
                else
                {
                    Console.WriteLine($"[WARNING] Stock {stockId} not found in close.parquet");
                    return null;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[ERROR] Failed to read close.parquet: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Calculates limit-up price. Delegates to TickSizeHelper.
        /// </summary>
        public static double CalculateLimitUp(double previousClose) => TickSizeHelper.CalculateLimitUp(previousClose);

        /// <summary>
        /// Calculates limit-down price. Delegates to TickSizeHelper.
        /// </summary>
        public static double CalculateLimitDown(double previousClose) => TickSizeHelper.CalculateLimitDown(previousClose);

        /// <summary>
        /// Gets the tick size as decimal for a given price.
        /// </summary>
        public static decimal GetTickSizeDecimal(double price) => TickSizeHelper.GetTickSizeDecimal(price);

        /// <summary>
        /// Adds placeholder computed columns to tick data (filled during backtest).
        /// </summary>
        public List<TickData> AddCalculatedColumns(List<TickData> data)
        {
            foreach (var row in data)
            {
                row.DayHighGrowthRate = 0.0;
                row.BidAvgVolume = 0.0;
                row.AskAvgVolume = 0.0;
                row.BalanceRatio = 0.0;
                row.DayHighBreakout = false;
            }
            return data;
        }

        /// <summary>
        /// Gets the company name for a stock ID.
        /// </summary>
        public string GetCompanyName(string stockId)
        {
            try
            {
                if (_companyNameCache == null)
                {
                    var possiblePaths = new[]
                    {
                        "company_basic_info.parquet",
                        @"D:\feature_data\company_basic_info.parquet",
                        @"C:\Users\User\Documents\_02_bt\Backtest_tick_module\company_basic_info.parquet"
                    };

                    foreach (var path in possiblePaths)
                    {
                        if (File.Exists(path))
                        {
                            _companyNameCache = ParquetHelper.ReadCompanyInfo(path);
                            Console.WriteLine($"[INFO] Loaded company info from: {path}");
                            break;
                        }
                    }

                    _companyNameCache ??= new Dictionary<string, string>();
                }

                return _companyNameCache.TryGetValue(stockId, out var name) ? name : stockId;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[WARNING] Failed to load company name: {ex.Message}");
                return stockId;
            }
        }

        /// <summary>
        /// Filters data to trading hours (09:00-13:30).
        /// </summary>
        public List<TickData> FilterTradingHours(List<TickData> data)
        {
            var start = new TimeSpan(9, 0, 0);
            var end = new TimeSpan(13, 30, 0);
            return data.Where(t => t.Time.TimeOfDay >= start && t.Time.TimeOfDay <= end).ToList();
        }

        /// <summary>
        /// Validates that data has required columns and is non-empty.
        /// </summary>
        public bool ValidateData(List<TickData> data)
        {
            if (data == null || data.Count == 0)
            {
                Console.WriteLine("[ERROR] Data is empty");
                return false;
            }

            // Check time ordering
            for (int i = 1; i < data.Count; i++)
            {
                if (data[i].Time < data[i - 1].Time)
                {
                    Console.WriteLine("[WARNING] Data not sorted by time, sorting now.");
                    data.Sort((a, b) => a.Time.CompareTo(b.Time));
                    break;
                }
            }

            return true;
        }

        // ===== Private helpers =====

        private static void SetOrderBookToNaN(TickData row)
        {
            // Replace 0 with NaN for order book columns (so they can be forward-filled)
            if (row.Bid1Volume == 0) row.Bid1Volume = double.NaN;
            if (row.Bid2Volume == 0) row.Bid2Volume = double.NaN;
            if (row.Bid3Volume == 0) row.Bid3Volume = double.NaN;
            if (row.Bid4Volume == 0) row.Bid4Volume = double.NaN;
            if (row.Bid5Volume == 0) row.Bid5Volume = double.NaN;
            if (row.Ask1Volume == 0) row.Ask1Volume = double.NaN;
            if (row.Ask2Volume == 0) row.Ask2Volume = double.NaN;
            if (row.Ask3Volume == 0) row.Ask3Volume = double.NaN;
            if (row.Ask4Volume == 0) row.Ask4Volume = double.NaN;
            if (row.Ask5Volume == 0) row.Ask5Volume = double.NaN;
            if (row.BidVolume5Level == 0) row.BidVolume5Level = double.NaN;
            if (row.AskVolume5Level == 0) row.AskVolume5Level = double.NaN;
            if (row.BidAskRatio == 0) row.BidAskRatio = double.NaN;
        }

        private static void ForwardFillOrderBook(List<TickData> data)
        {
            for (int i = 1; i < data.Count; i++)
            {
                var prev = data[i - 1];
                var curr = data[i];

                if (double.IsNaN(curr.Bid1Volume)) curr.Bid1Volume = prev.Bid1Volume;
                if (double.IsNaN(curr.Bid2Volume)) curr.Bid2Volume = prev.Bid2Volume;
                if (double.IsNaN(curr.Bid3Volume)) curr.Bid3Volume = prev.Bid3Volume;
                if (double.IsNaN(curr.Bid4Volume)) curr.Bid4Volume = prev.Bid4Volume;
                if (double.IsNaN(curr.Bid5Volume)) curr.Bid5Volume = prev.Bid5Volume;
                if (double.IsNaN(curr.Ask1Volume)) curr.Ask1Volume = prev.Ask1Volume;
                if (double.IsNaN(curr.Ask2Volume)) curr.Ask2Volume = prev.Ask2Volume;
                if (double.IsNaN(curr.Ask3Volume)) curr.Ask3Volume = prev.Ask3Volume;
                if (double.IsNaN(curr.Ask4Volume)) curr.Ask4Volume = prev.Ask4Volume;
                if (double.IsNaN(curr.Ask5Volume)) curr.Ask5Volume = prev.Ask5Volume;
                if (double.IsNaN(curr.BidVolume5Level)) curr.BidVolume5Level = prev.BidVolume5Level;
                if (double.IsNaN(curr.AskVolume5Level)) curr.AskVolume5Level = prev.AskVolume5Level;
                if (double.IsNaN(curr.BidAskRatio)) curr.BidAskRatio = prev.BidAskRatio;
            }
        }

        private static void EnsureFiveLevelTotals(TickData row)
        {
            // Compute bid_volume_5level if missing or NaN
            if (double.IsNaN(row.BidVolume5Level) || row.BidVolume5Level == 0)
            {
                double sum = 0;
                int count = 0;
                foreach (var v in new[] { row.Bid1Volume, row.Bid2Volume, row.Bid3Volume, row.Bid4Volume, row.Bid5Volume })
                {
                    if (!double.IsNaN(v)) { sum += v; count++; }
                }
                if (count > 0) row.BidVolume5Level = sum;
            }

            // Compute ask_volume_5level if missing or NaN
            if (double.IsNaN(row.AskVolume5Level) || row.AskVolume5Level == 0)
            {
                double sum = 0;
                int count = 0;
                foreach (var v in new[] { row.Ask1Volume, row.Ask2Volume, row.Ask3Volume, row.Ask4Volume, row.Ask5Volume })
                {
                    if (!double.IsNaN(v)) { sum += v; count++; }
                }
                if (count > 0) row.AskVolume5Level = sum;
            }
        }
    }

    /// <summary>
    /// Helper class for reading parquet files.
    /// In a real implementation, use Parquet.Net or Apache.Arrow NuGet packages.
    /// These methods provide the contract that actual parquet reading must fulfill.
    /// </summary>
    public static class ParquetHelper
    {
        /// <summary>
        /// Reads a feature parquet file and returns a list of TickData.
        /// Implementation requires Parquet.Net or Apache.Arrow.
        /// </summary>
        public static List<TickData> ReadParquetToTickData(string path)
        {
            // Placeholder: In production, use Parquet.Net / Apache.Arrow to read each column
            // and populate TickData objects. The parquet schema contains columns:
            // time, price, volume, tick_type, type, day_high, bid_ask_ratio,
            // bid1_volume..bid5_volume, ask1_volume..ask5_volume,
            // bid_volume_5level, ask_volume_5level, ratio_15s_300s,
            // pct_2min, pct_3min, pct_5min, low_1m, low_3m, low_5m, low_10m, low_15m
            Console.WriteLine($"[INFO] Reading parquet file: {path}");
            Console.WriteLine("[WARNING] ParquetHelper.ReadParquetToTickData requires Parquet.Net or Apache.Arrow implementation.");
            return new List<TickData>();
        }

        /// <summary>
        /// Reads close.parquet: returns Dictionary&lt;DateTime, Dictionary&lt;string, double&gt;&gt;
        /// where outer key is date, inner key is stock_id, value is close price.
        /// </summary>
        public static Dictionary<DateTime, Dictionary<string, double>> ReadCloseParquet(string path)
        {
            Console.WriteLine($"[INFO] Reading close parquet: {path}");
            Console.WriteLine("[WARNING] ParquetHelper.ReadCloseParquet requires Parquet.Net or Apache.Arrow implementation.");
            return new Dictionary<DateTime, Dictionary<string, double>>();
        }

        /// <summary>
        /// Reads company basic info parquet: returns Dictionary&lt;stockId, companyName&gt;.
        /// </summary>
        public static Dictionary<string, string> ReadCompanyInfo(string path)
        {
            Console.WriteLine($"[INFO] Reading company info: {path}");
            Console.WriteLine("[WARNING] ParquetHelper.ReadCompanyInfo requires Parquet.Net or Apache.Arrow implementation.");
            return new Dictionary<string, string>();
        }
    }
}
